File: scala/sbt/src/test/scala/io/gatling/grpc/demo/Feeders.scala
package io.gatling.grpc.demo

import java.util.concurrent.ThreadLocalRandom

import io.grpc._

object Feeders {

  val channelCredentials: Array[Map[String, ChannelCredentials]] =
    (1 to 3).map { i =>
      val credentials = TlsChannelCredentials
        .newBuilder()
        .keyManager(
          ClassLoader.getSystemResourceAsStream(s"certs/client$i.crt"),
          ClassLoader.getSystemResourceAsStream(s"certs/client$i.key")
        )
        .trustManager(ClassLoader.getSystemResourceAsStream("certs/ca.crt"))
        .build
      Map("channelCredentials" -> credentials)
    }.toArray

  private val Alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"

  private def random(alphabet: String, n: Int): String =
    LazyList.continually(ThreadLocalRandom.current.nextInt(alphabet.length)).map(alphabet).take(n).mkString

  private def randomString(n: Int): String = random(Alphabet, n)

  val randomNames: Iterator[Map[String, String]] =
    Iterator.continually {
      Map(
        "firstName" -> randomString(20),
        "lastName" -> randomString(20)
      )
    }
}

File: javascript/src/feeders.js
import { arrayFeeder } from "@gatling.io/core";

const availableChannelCredentials = [];
for (let i = 1; i <= 3; i++) {
  availableChannelCredentials.push({
    rootCerts: "certs/ca.crt",
    certChain: `certs/client${i}.crt`,
    privateKey: `certs/client${i}.key`
  });
}

export const channelCredentials = arrayFeeder(
  availableChannelCredentials.map((channelCredentials) => ({
    channelCredentials
  }))
);

const random = (alphabet, n) => {
  let s = "";
  for (let i = 0; i < n; i++) {
    const index = Math.floor(Math.random() * alphabet.length);
    s += alphabet.charAt(index);
  }
  return s;
};

const randomString = (n) => random("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ", n);

const randomNamesRecords = [];
for (let i = 0; i < 5; i++) {
  randomNamesRecords.push({
    firstName: randomString(20),
    lastName: randomString(20)
  });
}

export const randomNames = arrayFeeder(randomNamesRecords);
